src/Bicep.Cli/Services/CompilationService.cs

```diff
@@ -154,27 +152,30 @@ public async Task<(Uri, ImmutableDictionary<Uri, string>)> DecompileAsync(string
             return decompilation;
         }
 
-        private static ImmutableDictionary<BicepFile, ImmutableArray<IDiagnostic>> GetModuleRestoreDiagnosticsByBicepFile(SourceFileGrouping sourceFileGrouping, ImmutableHashSet<ModuleDeclarationSyntax> originalModulesToRestore, bool forceModulesRestore)
+        private static ImmutableDictionary<BicepFile, ImmutableArray<IDiagnostic>> GetModuleRestoreDiagnosticsByBicepFile(SourceFileGrouping sourceFileGrouping, ImmutableHashSet<ModuleSourceResolutionInfo> originalModulesToRestore, bool forceModulesRestore)
         {
-            static IEnumerable<IDiagnostic> GetModuleDiagnosticsPerFile(SourceFileGrouping grouping, BicepFile bicepFile, ImmutableHashSet<ModuleDeclarationSyntax> originalModulesToRestore, bool forceModulesRestore)
+            static IEnumerable<IDiagnostic> GetModuleDiagnosticsPerFile(SourceFileGrouping grouping, BicepFile bicepFile)
             {
                 foreach (var module in bicepFile.ProgramSyntax.Declarations.OfType<ModuleDeclarationSyntax>())
                 {
-                    if (!forceModulesRestore && !originalModulesToRestore.Contains(module))
-                    {
-                        continue;
-                    }
-
                     if (grouping.TryGetErrorDiagnostic(module) is {} errorBuilder)
                     {
                         yield return errorBuilder(DiagnosticBuilder.ForPosition(module.Path));
                     }
                 }
             }
 
-            return sourceFileGrouping.SourceFiles
-                .OfType<BicepFile>()
-                .ToImmutableDictionary(bicepFile => bicepFile, bicepFile => GetModuleDiagnosticsPerFile(sourceFileGrouping, bicepFile, originalModulesToRestore, forceModulesRestore).ToImmutableArray());
+            if (forceModulesRestore)
+            {
+                return sourceFileGrouping.SourceFiles
+                    .OfType<BicepFile>()
+                    .ToImmutableDictionary(bicepFile => bicepFile, bicepFile => GetModuleDiagnosticsPerFile(sourceFileGrouping, bicepFile).ToImmutableArray());
+            }
+
+            return originalModulesToRestore.SelectMany(t => t.ParentTemplateFile is BicepFile bicepFile ? new [] { (bicepFile, t.ModuleDeclaration) } : Enumerable.Empty<(BicepFile, ModuleDeclarationSyntax)>())
+                .SelectMany(t => sourceFileGrouping.TryGetErrorDiagnostic(t.Item2)?.Invoke(DiagnosticBuilder.ForPosition(t.Item2.Path)) is IDiagnostic diagnostic ? new [] { (t.Item1, diagnostic) } : Enumerable.Empty<(BicepFile, IDiagnostic)>())
+                .ToLookup(t => t.Item1, t => t.Item2)
+                .ToImmutableDictionary(g => g.Key, g => g.ToImmutableArray());
```

Comment:
It would be nice to retain the common logic rather than duplicating it, and I also feel like the Linq is getting a bit hard to follow. What about something like the following:

```chsarp
private static ImmutableDictionary<BicepFile, ImmutableArray<IDiagnostic>> GetModuleRestoreDiagnosticsByBicepFile(SourceFileGrouping sourceFileGrouping, ImmutableHashSet<ModuleSourceResolutionInfo> originalModulesToRestore, bool forceModulesRestore)
{
    static IEnumerable<ModuleSourceResolutionInfo> GetAllModulesForRestore(SourceFileGrouping grouping)
    {
        foreach (var file in grouping.SourceFiles.OfType<BicepFile>())
        {
            foreach (var module in file.ProgramSyntax.Declarations.OfType<ModuleDeclarationSyntax>())
            {
                yield return new(module, file);
            }
        }
    }

    static IEnumerable<(BicepFile, IDiagnostic)> GetModuleDiagnosticsPerBicepFile(SourceFileGrouping grouping, IEnumerable<ModuleSourceResolutionInfo> modulesToRestore)
    {
        foreach (var (module, file) in modulesToRestore)
        {
            if (file is BicepFile bicepFile &&
                grouping.TryGetErrorDiagnostic(module) is {} errorBuilder)
            {
                yield return (bicepFile, errorBuilder(DiagnosticBuilder.ForPosition(module.Path)));
            }
        }
    }

    var modulesToRestore = forceModulesRestore ? GetAllModulesForRestore(sourceFileGrouping) : originalModulesToRestore;

    return GetModuleDiagnosticsPerBicepFile(sourceFileGrouping, modulesToRestore)
        .ToLookup(t => t.Item1, t => t.Item2)
        .ToImmutableDictionary(g => g.Key, g => g.ToImmutableArray());
}
```